### PR TITLE
Get rid of 30s ResyncPeriod in endpoint controller

### PR DESCRIPTION
**Pull request by shyamjvs**
Ref: #47597 
This should fix one of the demons of endpoint controller.

/cc @smarterclayton @gmarek

### CODE REVIEW

**Comment by shyamjvs** [comment]
cc @jsravn

File: pkg/controller/endpoint/endpoints_controller.go [hunk 1]
```diff
@@ -231,14 +222,19 @@ func (e *EndpointController) deletePod(obj interface{}) {
 		e.addPod(obj)
 		return
 	}
-	podKey, err := keyFunc(obj)
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", obj, err))
+	// If we reached here it means the pod was deleted but it's final state is unrecorded.
```
Comment (seh) [inline pkg/controller/endpoint/endpoints_controller.go]: - [ ] s/it's/its/

File: pkg/controller/endpoint/endpoints_controller.go [hunk 2]
```diff
@@ -231,14 +222,19 @@ func (e *EndpointController) deletePod(obj interface{}) {
...
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get object from tombstone %#v", obj))
 		return
 	}
-	glog.V(4).Infof("Pod %q was deleted but we don't have a record of its final state, so it will take up to %v before it will be removed from all endpoint records.", podKey, FullServiceResyncPeriod)
-
-	// TODO: keep a map of pods to services to handle this condition.
+	pod, ok := tombstone.Obj.(*v1.Pod)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("Tombstone contained object that is not a Pod %#v", obj))
```
Comment (seh) [inline pkg/controller/endpoint/endpoints_controller.go]: - [ ] s/a Pod /a Pod: /  
  (Add an intervening colon.)

**Review by seh — CHANGES_REQUESTED** [review]

File: pkg/controller/endpoint/endpoints_controller.go (hunk 1 above)
Comment (shyamjvs) [inline pkg/controller/endpoint/endpoints_controller.go]: Done.

File: pkg/controller/endpoint/endpoints_controller.go (hunk 2 above)
Comment (shyamjvs) [inline pkg/controller/endpoint/endpoints_controller.go]: Done.

**Comment by shyamjvs** [comment]
/test pull-kubernetes-bazel
(seems unrelated)

**Review by seh — APPROVED** [review]
Thanks for addressing those nits.

It's not clear to me how to remove my earlier objection in a GitHub review without approving the change. I defer to @smarterclayton and @gmarek to review the change as well.

**Comment by shyamjvs** [comment]
I'm adding v1.7 milestone. Feel free to remove if you have a good reason.

**Comment by jsravn** [comment]
@shyamjvs I think with moving to using the endpoints informer, the resync is much less of a problem. It will basically just be CPU overhead now.

**Comment by smarterclayton** [comment]
A proper informer driven controller should never need to resync unless it
deals with external "non-informer" data.  So any change to use informers
should work without a 30s resync, and we should increase the resync to
ensure we aren't broken without it.

On Jun 20, 2017, at 11:57 AM, Shyam JVS <notifications@github.com> wrote:

I'm adding v1.7 milestone. Feel free to remove if you have a good reason.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/47788#issuecomment-309804729>,
or mute the thread

**Comment by seh** [comment]
@smarterclayton, for a controller that creates Kubernetes objects, should it react to changes in those same objects it creates? Ideally we'd like to detect when some other agent tampered with the objects this controller creates (and maintains) and correct them if they differ from our intended state, but then we'll wind up reacting to this controller's own changes too. Is there a way around that reacting to and inspecting of your own induced changes, or do you consider that to be a worthwhile tax that's too hard to avoid?

I ask because I'm working on a controller of my own that faces this same conundrum. It creates, modifies, and deletes _ResourceQuota_ objects, and those objects change frequently (the admission control plugin changes the status, at least) as other objects get created and deleted. If my controller watches _ResourceQuota_ objects, I'll be reacting both to my own controller's changes and those to the status, neither of which I care about. What do you say there? Watch them, or don't watch them?

**Comment by smarterclayton** [comment]
I think this is important, but not quite important enough to risk release over.  We could consider a cherry-pick post release if this goes smoothly in master and we get good soak time on it.

@kubernetes/release-reviewers

**Comment by shyamjvs** [comment]
@liggitt @wojtek-t fyi

**Comment by mikedanese** [comment]
Get rid of it entirely? Why not just make it really big?

**Comment by gmarek** [comment]
Get rid entirely. We don't have it anywhere else (or rather we should not have it anywhere else).

**Comment by jsravn** [comment]
Isn't there a race condition in the way the endpoint controller handles items on the work queue that could cause missed updates?

1. endpoint update A arrives on workqueue
2. worker wakes up, processes item
3. worker almost finishes syncing service
4. endpoint update B arrives on workqueue
5. worker finishes processing A, removes key from workqueue
6. endpoint update B is missed

In this case if we remove the resync we could miss endpoint updates entirely. That could be catastrophic for production services. I think removing the resync is high risk unless we're sure all data races are fixed. But it's hard to get right (see the attach_detach_controller for example, which has just recently added resync logic).

**Comment by kargakis** [comment]
Doesn't the attach-dettach controller deal with external to k8s resources? Clayton's [comment](https://github.com/kubernetes/kubernetes/pull/47788#issuecomment-309922888) applies in that case.

**Comment by mikedanese** [comment]
@jsravn the workqueue ack checks if the object is dirty and readds it to the queue if it is.

**Comment by jsravn** [comment]
Yeah, ignore me, I read the work queue and saw the dirty set which covers that case.

**Comment by mikedanese** [comment]
Can someone explain to me how we are not in violation of: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/principles.md#control-logic point one if we don't occasionally resync/relist? Do we not care about that anymore?

**Comment by gmarek** [comment]
That's a good question - I believe we stopped treating Informers as 'cache'. @kubernetes/sig-api-machinery-misc

**Comment by jsravn** [comment]
@kargakis I'm mostly familiar with the aws implementation of it. But I believe the attach_detach_controller deals exclusively in k8s resources. It's kubelet that updates those resources based on external state, like whether the volume is mounted locally etc. The known data races in the attach detach controller have been purely code bugs, but there have been so many of them due to complexity that the resync was added as a catch all.

**Comment by shyamjvs** [comment]
Just to clarify, we still have a resync period but it's quite high (~21 hrs
iirc). AddResourceHandler internally calls
AddResourceHandlerWithResyncPeriod, but with this default high value.

**Comment by shyamjvs** [comment]
Correction: AddEventHandler
Ref: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L265

**Comment by shyamjvs** [comment]
@mikedanese Could you approve the PR? Thanks.

**Pull request merged**